Jenkinsfile

```diff
@@ -0,0 +1,307 @@
+    // If the PR does not pass the static checks,
+    // then the rest of the CI checks will be skipped by Jenkins.
+    stage('Static checks') {
+      agent { label "master" }
+      steps {
+        script {
+          echo "In static check"
+        }
+      }
+    }
+
+//    stage('examine the rootdir') {
+//      agent { label "master" }
+//      steps {
+//        script {
+//          sh "pwd"
+//          sh "ls"
+//        }
+//      }
+//    }
+
+    // Load up our test matrix data from the YAML file so we can tell
+    // which tests are to be run on which distro/jobs
+    stage('Load test matrix data') {
+      agent { label "master" }
+      steps {
+// FIXME - if we are test the CI repo, then we need to pick the config file up from the
+// Jenkins checked out PR (root of WORKSPACE), and not the golang checked out
+// ci repo. Code that up when we have access to the ghprbGhRepository var...
+// FIXME
+// // By default we run the tests using the git checkout out test repo
+// testcode_rootdir = env.ci_repo_dir
+// if ( env.ghprbGhRepository != null ) {
+//   // but, if this PR is from the tests repo, use the Jenkins checked out
+//   // source/branch of that instead, in case the PR is updating the test code.
+//   if ( env.ghprbGhRepository == env.ci_repo_name ) {
+//     testcode_rootdir = ""
+//   }
+// }
+        //dir("${testcode_rootdir}") {
+          script {
+            sh "pwd"
+            sh "ls"
+            testmatrix = readYaml file: "${testmatrix_file}"
+            echo "testmatrix looks like ${testmatrix}"
+         // }
+        }
+      }
+    }
+
+    // Run the CI on the primary distro (note, that does not mean this is our
+    // favoured distro, it just means we needed to pick one as the primary
+    // smoke test). If the primary distro does not pass, the rest of the distro
+    // tests will be skipped by Jenkins.
+    stage('Ubuntu18.04') {
+      when {
+        expression { SKIPTESTS != 'true' }
+      }
+      agent { label "master" }
+      steps {
+        echo "Am running ${STAGE_NAME}"
+        testrunner ( STAGE_NAME )
+      }
+    }
+
+    // If we have passed all previous steps, finally we can parallel run the
+    // rest of the distro checks.
+    stage('Run main jobs') {
+      when {
+        expression { SKIPTESTS != 'true' }
+      }
+      parallel {
+        stage('Fedora') {
+          agent { label "master" }
+          steps {
+            echo "Am running ${STAGE_NAME}"
+            testrunner ( STAGE_NAME )
+          }
+        }
+        stage('Centos') {
+          agent { label "master" }
+          steps {
+            echo "Am running ${STAGE_NAME}"
+            testrunner ( STAGE_NAME )
+          }
+        }
+      }
+    }
+  }
+}
+
+
+// Run our tests on the 'distroName' distro provided.
+// Use that name to look in the testmatrix to work out which tests
+// are to be run on this repo/distro combination.
+
+def testrunner(distroName) {
```

Comment:
Question: aiui, all `stage`'s are run and if any fail, no further stage is run. So how does a function like this signify failure? What does a "bare `return` mean" (presumably success)?

Reply:
yes, a return is success. Right now I had to hack a little bit here as `ghprbGhRepository` is not set on the test servers when running this - and I wanted to run it to be able to grab the screenshot. Ultimately we'll make it:
- more robust
- able to run on master merge (non-PR) builds as well I expect